docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+}
+draw(10, 20)
+draw(20, 20, true)
+```
+
+When writing Melange bindings, we can add labeled arguments to make things more
+clear:
+
+```ocaml
+external draw : x:int -> y:int -> ?border:bool -> unit -> unit = "draw"
+  [@@module "MyGame"]
+
+let () = draw ~x:10 ~y:20 ~border:true ()
+let () = draw ~x:10 ~y:20 ()
+```
+
+Generates:
+
+```js
+var MyGame = require("MyGame");
+
+MyGame.draw(10, 20, true);
+MyGame.draw(10, 20, undefined);
+```
+
+The generated JavaScript function is the same, but now the usage in Melange is
+much clearer.
+
+**Note**: in this particular case, a final param of type unit, `()` must be
+added after `border`, since `border` is an optional argument at the last
+position. Not having the last param `unit` would lead to a warning, which is
+explained in detail [in the OCaml
+documentation](https://ocaml.org/docs/labels#warning-this-optional-argument-cannot-be-erased).
+
+Note that you can freely reorder the labels on the Melange side, they will
+appear in the right order on the JavaScript output:
+
+```ocaml
+external draw : x:int -> y:int -> ?border:bool -> unit -> unit = "draw"
+  [@@module "MyGame"]
+let () = draw ~x:10 ~y:20 ()
+let () = draw ~y:20 ~x:10 ()
+```
+
+Generates:
+
+```js
+var MyGame = require("MyGame");
+
+MyGame.draw(10, 20, undefined);
+MyGame.draw(10, 20, undefined);
+```
+
+#### Calling a function as an object property
+
+If we need to call a function that happens to be set as a JavaScript object
+property, Melange provides the attribute `bs.send` so we can define it with just
+one `external`:
+
+```ocaml
+(* Abstract type for the `document` global *)
+type document
+
+external document : document = "document" [@@bs.val]
+external get_by_id : document -> string -> Dom.element = "getElementById"
+  [@@bs.send]
+
+let el = get_by_id document "my-id"
+```
+
+```js
+var el = document.getElementById("my-id");
+```
+
+When using `bs.send`, the first argument will be the object that holds the
+property with the function we want to call. In the example above, we are binding
+to the JavaScript object `document` using an abstract type, a very useful
+technique for bindings but also in other use cases. The "Encapsulation" section
+[in the OCaml Cornell
+textbook](https://cs3110.github.io/textbook/chapters/modules/encapsulation.html)
+provides a great overview to understand what abstract types are and when they
+are useful.
+
+##### Chaining
+
+It is common to find this kind of API in JavaScript: `foo().bar().baz()`. This
+kind of API can be designed with Melange externals and the `bs.send` attribute,
+in combination with the [the pipe operator](todo-fix-me.md).
+
+#### Variadic function arguments
+
+Sometimes JavaScript functions take an arbitrary amount of arguments. For this
+cases, Melange provides the `bs.variadic` attribute, which can be attached to
+the `external` declaration. However, there is one caveat: all the variadic
+arguments need to belong to the same type.
```

Comment:
should we document the solution for this?

```ocaml
type hide = Hide : 'a -> hide [@@unboxed]

external join : hide array -> string = "join" [@@bs.module "path"] [@@bs.splice]

let v = join [| Hide "a"; Hide 2 |]

```

which generates:

```js
// Generated by Melange
'use strict';

var Path = require("path");

var v = Path.join("a", 2);

exports.v = v;
/* v Not a pure module */
```

The same trick can be used for `Js.Dict`s

Reply:
Added solution for variadic in https://github.com/melange-re/documentation-site/pull/12/commits/b3bad3de602c738f17111b0e1689b0f3534ea247.

> The same trick can be used for `Js.Dict`s

The `unboxed` solution with existential types (is that what `hide` is?) for dictionaries would mean values can be added to the dictionary, but not be read from it, right?

Reply:
yup, section looks good.